### PR TITLE
Add `Effect.liftPredicate`

### PR DESCRIPTION
**Pull request by LaureRC**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Add `Effect.liftPredicate` (similar to recently merged `Either.liftPredicate`: https://github.com/Effect-TS/effect/pull/2919). This is a proposition though, I'm not 100% sure it's really needed, as we might be able to use `Either.liftPredicate` (at least that's the case for the use cases I was looking into in my current project). So I'll keep the PR as a draft for now to see if I find some good use cases where it's needed, otherwise I'll close the PR. Happy to get some feedback about the idea though!

## Related

N/A

### CODE REVIEW

**Comment by tim-smart** [comment]
I think the Effect.filter* apis are enough:

```ts
Effect.succeed(0).pipe(
  Effect.filterOrFail(...)
)
```

I'll let someone else weigh in before we decide to continue / close this PR.

**Comment by LaureRC** [comment]
Thank you for your reviews!

> I think the Effect.filter* apis are enough:

I understand your point, I thought it would be more convenient to directly have this `liftPredicate`, rather than have to wrap the initial value in an `Effect.succeed` to be able to use the `Effect.filter*` methods

> That's right, this is literally Either.liftPredicate typed differently:

Right, but then does that mean you don't recommend adding it for Effect? We can use `Either.liftPredicate` which is going to be released in the next minor anyway so I'm happy with that 🙂 but apparently some people do not use `Either` that much, so it could be convenient for them to not have to import (and know) about another module than `Effect`. I'm aware this is not a super strong argument though, so if you think that's not interesting enough I won't be mad if we close this subject 😄

**Comment by gcanti** [comment]
Actually, I'm not sure they are *literally* the same,`Effect.succeed`/`Effect.fail` and `Either.right`/`Either.left` return different values.

**Comment by LaureRC** [comment]
> Actually, I'm not sure they are literally the same,Effect.succeed/Effect.fail and Either.right/Either.left return different values.

Indeed, in some cases I identified in my current project, it does not change much to use the Either one. As I said before, I was pushing a little bit for this `Effect.liftPredicate` as I still think it could have value to have this method in both `Either` and `Effect` modules, to avoid having to "switch context" and use the `Either` module if not really needed. 
So I'll reopen the PR as I still think it could be useful, but don't hesitate to push back on it if you disagree 🙂 (as I didn't get an explicit no-go for now I'll insist a little bit more 😄)

File: packages/effect/src/Effect.ts [hunk 1]
```diff
@@ -2159,6 +2159,61 @@ export const uninterruptibleMask: <A, E, R>(
...
+    self: A,
+    refinement: Refinement<A, B>,
+    orFailWith: (a: A) => E
+  ): Effect<B, E>
+  <A, E>(
+    self: A,
+    predicate: Predicate<NoInfer<A>>,
+    orFailWith: (a: NoInfer<A>) => E
+  ): Effect<A, E>
+} = dual(
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:2211]: If you can move the implementation into `internal/core-effect.ts`, to be consistent with the rest of the apis.

File: packages/effect/src/Effect.ts [hunk 2]
```diff
@@ -2159,6 +2159,61 @@ export const uninterruptibleMask: <A, E, R>(
...
+  3,
+  <A, E>(a: A, predicate: Predicate<A>, orFailWith: (a: A) => E): Effect<A, E> =>
+    predicate(a) ? succeed(a) : fail(orFailWith(a))
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:2214]: This needs to be wrapped in `Effect.suspend`, in case the predicate throws an error.

**Review by tim-smart — CHANGES_REQUESTED** [review]

**Comment by tim-smart** [comment]
I'll make the changes now to get this into the next minor release.

**Review by tim-smart — APPROVED** [review]

**Comment by LaureRC** [comment]
> I'll make the changes now to get this into the next minor release.

Thank you for handling this! I'm glad to see this released 😊

**Pull request merged**